### PR TITLE
*_scatter ops should preserve input stride/storage_offset

### PR DESCRIPTION
**Pull request by bdhirsh**
It turns out that we *do* need to update *_scatter ops to return the exact same strides as their inputs. I added a test to `test/test_functionalization.py`, which now trips thanks to Ed's functionalization stride debugging check. It only actually ends up tripping silent correctness if you try to .backward() on that function.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89532
* #90870
* __->__ #91029

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
test would be nice

**Comment by bdhirsh** [comment]
Done. I also had to update a ton of expect tests: eagerly applying a mutation back to the base during functionalization was also needed to get that test passing, and causes our graphs to wobble.

**Comment by bdhirsh** [comment]
This is also (unfortunately) technically BC-breaking, although hopefully:
(1) very few people are actually using these *_scatter ops
(2) even fewer people rely on the old behavior of `*_scatter(non_contiguous_input)` -> contiguous output

File: aten/src/ATen/functorch/BatchRulesScatterOps.cpp [hunk 1]
```diff
@@ -1074,6 +1075,12 @@ TORCH_LIBRARY_IMPL(aten, FuncTorchBatched, m) {
   VMAP_SUPPORT(scatter_add, scatter_add_batch_rule);
   VMAP_SUPPORT2(scatter, reduce, scatter_reduce_batch_rule);
   VMAP_SUPPORT2(scatter, value_reduce, scatter_value_reduce_batch_rule);
+  // as_strided_scatter does not work with the for-loop fallback today,
+  // because as_strided_scatter will return an output that matches
+  // the strides/storage_offset of its input.
+  // With the for loop fallback, each input tensor is a slice into
+  // the larger batched tensor.
+  m.impl("as_strided_scatter", torch::CppFunction::makeFromBoxedFunction<&vmapErrorFallback>());
```
Comment (bdhirsh) [inline aten/src/ATen/functorch/BatchRulesScatterOps.cpp:1083]: cc @zou3519, I ended up adding an error fallback for vmap `as_strided_scatter`, and xfail'ing some more tests

Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesScatterOps.cpp:1083]: LGTM

**Pull request closed without merging**